File: solutions/052.N-Queens_II/AC_dfs_n!.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n!.java
*  Create Date: 2015-02-03 13:16:15
*  Descripton:   
*/

import java.util.Scanner;

public class Solution {

    public int totalNQueens(int n) {
        cnt = 0;
        col = new int[n];
        m_diag = new int[2 * n];
        a_diag = new int[2 * n];
        N = n;

        dfs(0);

        return cnt;
    }

    private int cnt;
    private int[] col;
    private int[] m_diag;
    private int[] a_diag;
    private int N;

    private void dfs(int row) {
        if (row == N) {
            ++cnt;
            return;
        }

        for (int i = 0; i < N; ++i) {
            if (col[i] != 0 || m_diag[row + i] != 0 || a_diag[row - i + N] != 0)
                continue;
            col[i] = m_diag[row + i] = a_diag[row - i + N] =  1;
            dfs(row + 1);
            col[i] = m_diag[row + i] = a_diag[row - i + N] =  0;

        }
    }


    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        System.out.println(s.totalNQueens(4));
    }

}

File: solutions/052.N-Queens_II/AC_dfs_n!.py
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_n!.py
# Create Date: 2015-02-03 11:55:43
# Descripton:   


class Solution:
    def dfs(self, row):
        if row == self.n:
            self.cnt += 1
            return
        for i in range(self.n):
            if self.col[i] != 0 or self.m_diag[row + i] != 0 or self.a_diag[row - i + self.n] != 0:
                continue
            self.col[i] = 1
            self.m_diag[row + i] = 1
            self.a_diag[row - i + self.n] = 1
            self.dfs(row + 1)
            self.col[i] = 0
            self.m_diag[row + i] = 0
            self.a_diag[row - i + self.n] = 0

    # @return a list of lists of string
    def totalNQueens(self, n):
        self.cnt = 0
        self.col = [0] * n
        self.m_diag = [0] * (2 * n)
        self.a_diag = [0] * (2 * n)
        self.n = n
        self.dfs(0)
        return self.cnt



s = Solution()
print s.totalNQueens(4)



